gdal/ogr/ogrsf_frmts/cad/ogrcadlayer.cpp

```diff
@@ -0,0 +1,639 @@
+
+OGRFeature *OGRCADLayer::GetNextFeature()
+{
+    OGRFeature *poFeature = GetFeature( nNextFID );
+    ++nNextFID;
+
+    if( poFeature == NULL )
+        return( NULL );
+
+    if( ( m_poFilterGeom == NULL ||  FilterGeometry( poFeature->GetGeometryRef() ) )
+        && ( m_poAttrQuery == NULL || m_poAttrQuery->Evaluate( poFeature ) ) )
+    {
+        return poFeature;
+    }
+}
+
+OGRFeature *OGRCADLayer::GetFeature( GIntBig nFID )
+{
+    if( poCADLayer.getGeometryCount() <= nFID
+        || nFID < 0 )
+    {
+        return NULL;
+    }
+
+    OGRFeature  *poFeature = NULL;
+    CADGeometry *poCADGeometry = poCADLayer.getGeometry( nFID );
+
+    if( NULL == poCADGeometry || GetLastErrorCode() != CADErrorCodes::SUCCESS )
+    {
+        CPLError( CE_Failure, CPLE_NotSupported,
+                 "Failed to get geometry with ID = %lld from layer \"%s\". Libopencad errorcode: %d",
+                 nFID, poCADLayer.getName().c_str(), GetLastErrorCode() );
+        return NULL;
+    }
+
+    poFeature = new OGRFeature( poFeatureDefn );
+    poFeature->SetFID( nFID );
+    poFeature->SetField( "thickness", poCADGeometry->getThickness() );
+
+    if( poCADGeometry->getEED().size() != 0 )
+    {
+        std::string sEEDAsOneString = "";
+        for ( std::vector<std::string>::const_iterator
+              iter = poCADGeometry->getEED().cbegin();
+              iter != poCADGeometry->getEED().cend(); ++iter )
+        {
+            sEEDAsOneString += *iter;
+            sEEDAsOneString += ' ';
+        }
+
+        poFeature->SetField( "extentity_data", sEEDAsOneString.c_str() );
+    }
+
+    RGBColor stRGB = poCADGeometry->getColor();
+    GIntBig adRGB[3] { stRGB.R, stRGB.G, stRGB.B };
+    poFeature->SetField( "color", 3, adRGB);
+
+    std::stringstream oStringStream;
+    oStringStream << "PEN(c:#" << std::hex << adRGB[0] << adRGB[1] << adRGB[2];
+    oStringStream << ",w:5px)" << std::dec;
+    poFeature->SetStyleString( oStringStream.str().c_str() );
+
+    std::vector< CADAttrib > oBlockAttrs = poCADGeometry->getBlockAttributes();
+    for( std::vector< CADAttrib >::const_iterator citerBlockAttrs = oBlockAttrs.cbegin(); 
+         citerBlockAttrs != oBlockAttrs.cend(); ++citerBlockAttrs )
+    {
+        for( std::vector< std::string >::const_iterator citerFeatAttrs = asFeaturesAttributes.cbegin(); 
+            citerFeatAttrs != asFeaturesAttributes.cend(); 
+            ++citerFeatAttrs)
+        {
+            if( citerBlockAttrs->getTag() == *citerFeatAttrs )
+            {
+                poFeature->SetField( (*citerFeatAttrs).c_str(), citerBlockAttrs->getTextValue().c_str() );
+            }
+        }
+    }
+
+    switch( poCADGeometry->getType() )
+    {
+        case CADGeometry::POINT:
+        {
+            CADPoint3D * const poCADPoint = ( CADPoint3D* ) poCADGeometry;
+            CADVector stPositionVector = poCADPoint->getPosition();
+
+            poFeature->SetGeometryDirectly( new OGRPoint( stPositionVector.getX(),
```

Comment:
You should attach the spatial reference to the geometry with OGRGeometry::assignSpatialReference(), otherwise test_ogrsf will complain. True for all cases below